daemon/src/online_status_actor.rs

```diff
@@ -0,0 +1,105 @@
+use std::time::Duration;
+
+use async_trait::async_trait;
+use libp2p_core::PeerId;
+use tokio::sync::watch;
+use xtra::prelude::*;
+use xtra_libp2p::endpoint;
+use xtra_libp2p::Endpoint;
+use xtra_libp2p::GetConnectionStats;
+use xtra_productivity::xtra_productivity;
+
+use crate::connection::ConnectionStatus;
+
+/// Actor that transmits updates of ConnectionStatus of a specified PeerId based on
+/// information transmitted by the Endpoint via a watch channel.
+pub struct Actor {
+    endpoint: Address<Endpoint>,
+    watched_peer: PeerId,
+    sender: watch::Sender<ConnectionStatus>,
+}
+
+impl Actor {
+    pub fn new(
+        endpoint: Address<Endpoint>,
+        watched_peer: PeerId,
+        sender: watch::Sender<ConnectionStatus>,
+    ) -> Self {
+        Self {
+            endpoint,
+            watched_peer,
+            sender,
+        }
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    async fn started(&mut self, ctx: &mut Context<Self>) {
+        tracing::debug!(
+            "Online status watch actor started. Monitoring for peer id changes: {:?}",
+            self.watched_peer
+        );
+
+        match self.endpoint.send(GetConnectionStats).await {
+            Ok(connection_stats) => {
+                let status = if connection_stats
+                    .connected_peers
+                    .contains(&self.watched_peer)
+                {
+                    ConnectionStatus::Online
+                } else {
+                    ConnectionStatus::Offline { reason: None }
+                };
+                self.sender
+                    .send(status)
+                    .expect("Receiver to outlive this actor");
+            }
+            Err(e) => {
+                tracing::error!(
+                    "Unable to receive connection stats from the endpoint upon startup: {e:#}"
+                );
+                // This code path should not be hit, but in case we run into an error this sleep
+                // prevents a continuous endless loup of restarts.
+                self.sender
+                    .send(ConnectionStatus::Offline { reason: None })
+                    .expect("Receiver to outlive this actor");
+                tokio::time::sleep(Duration::from_secs(2)).await;
+
+                ctx.stop();
+            }
+        }
+    }
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+#[xtra_productivity(message_impl = false)]
+impl Actor {
+    async fn handle_connection_established(&mut self, msg: endpoint::ConnectionEstablished) {
+        tracing::debug!(
+            "Adding newly established connection to online_status: {:?}",
+            msg.peer
+        );
+        if msg.peer == self.watched_peer {
+            self.sender
+                .send(ConnectionStatus::Online)
+                .expect("Receiver to outlive this actor");
+        }
+    }
+
+    async fn handle_connection_dropped(&mut self, msg: endpoint::ConnectionDropped) {
```

Comment:
If the connection was dropped, is there a reason why this actor should exist? Might be better to shut it down?

Reply:
I meant: 
There will be one actor per connected peer, even if the peer only came online once and never ever again. Would it not be better to just drop the actor as well if the peer disappears and just recreate it when it appears again.

I don't think it's a big issue as the memory footprint of this actor is quite small.

Reply:
ah, I see what you're saying! this actor actually is taker-only, so there only over will be one of them; the peer id is configured at startup. (for some reason, taker-specific code still lives in the `daemon` crate, we should clean this up :/)

Reply:
ah, got it. Thanks for the clarification :)